include/deal.II/base/mpi.templates.h

```diff
@@ -264,6 +265,29 @@ namespace Utilities
 
 
 
+    template <typename Number>
+    void
+    sum (const SparseMatrix<Number> &local,
+         const MPI_Comm &mpi_communicator,
+         SparseMatrix<Number> &global)
+    {
+      Assert(local.get_sparsity_pattern() == global.get_sparsity_pattern(),
+             ExcMessage("The sparsity pattern of the local and the global matrices should match."));
+#ifdef DEAL_II_WITH_MPI
+      // makes use of the fact that the matrix stores its data in a
+      // contiguous array.
+      sum(ArrayView<const Number> (&local.val[0], local.n_nonzero_elements()),
+          mpi_communicator,
+          ArrayView<Number>(&global.val[0], global.n_nonzero_elements()));
```

Comment:
Can you add an assertion here that checks that all processors have the same value for `n_nonzero_elements()` by means of `MPI::min()` and `MPI::max()` in debug mode? Or, even better, move it into the implementation inside the `MPI::sum` function with the `ArrayView` argument? I've checked there and it seems it is not checked there. It might be that the MPI implementation also checks the consistency of the sizes, but I think we would have much better error messages at least.

This assertion would give some basic safety of consistent state over the processors, even though it's of course not a guarantee.

Reply:
done.

Reply:
now actually done this properly, had to do things manually, though.